File: WorkerNode/my-app/src/main/java/com/mycompany/app/WorkerNode.java
package com.mycompany.app;

import java.io.IOException;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingDeque;

import org.apache.pulsar.client.api.Consumer;
import org.apache.pulsar.client.api.Message;
import org.apache.pulsar.client.api.Producer;
import org.apache.pulsar.client.api.PulsarClient;
import org.apache.pulsar.client.api.SubscriptionType;

// 工作节点
public class WorkerNode
{
    // 参数
    static String nodeType;
    static String apiURL;
    static String adminAddress;
    static String pulsarURL;
    static String pulsarToken;
    static String topicName;
    static String subscriptionName;
    static int maxProcessNum;
    static String apiKey;
    static Boolean debug;
    static String servicePulsarURL;
    static String servicePulsarToken;
    static BlockingQueue<Message <byte []>> queue;

    // 手动初始化 测试用
    static void initForTest()
    {
        nodeType = "Api";
        adminAddress = "";
        pulsarURL = "pulsar://localhost:6650";
        topicName = "my-topic";
        subscriptionName = "my-subscription";
        maxProcessNum = 10;
        apiURL = "https://api.openai-hk.com/v1/chat/completions";
        apiKey = "hk-j9e9al1000037138f0cd6a31058a83dbb7a63f56fd48788c";
        queue = new LinkedBlockingDeque<>();
    }

    // 从环境变量读取初始化
    void init()
    {
        
    }

    // 运行
    public static void main( String[] args ) throws Exception
    {
        WorkerNode.initForTest();
        // 创建 pulsar 客户端
        PulsarClient client = PulsarClient.builder()
        .serviceUrl(pulsarURL)
        .build();
        
        // 创建生产者生产消息
        Producer<byte[]> producer = client.newProducer()
        .topic(topicName)
        .create();
        producer.send("this is a message from java".getBytes());

        // 创建消费者接收消息
        Consumer<byte []> consumer = client.newConsumer().topic(topicName)
            .subscriptionName(subscriptionName)
            .subscriptionType(SubscriptionType.Shared)
            .subscribe();

        // 创建处理消息的处理器
        Processor[] processors = new Processor[maxProcessNum];
        for(int i=0;i<maxProcessNum;i++)
        {
            processors[i] = new Processor(nodeType, queue, consumer,apiURL,apiKey,producer);
            processors[i].start("Thread " + String.valueOf(i));
        }

        Thread.sleep(500);

        // 不断接收消息
        while(true)
        {
            Message<byte []> msg = consumer.receive();
            queue.put(msg);
        }
    }
}

// 消费者，接收并处理消息
class Processor implements Runnable
{
    // 参数
    Thread th;
    String nodeType;
    String apiURL;
    String apiKey;
    Producer<byte []> producer;
    BlockingQueue<Message <byte []>> queue;
    Consumer<byte []> consumer;
    
    // 构造函数
    Processor(String nodeType,BlockingQueue<Message <byte []>> queue,Consumer<byte []> consumer,String apiURL,String apiKey,Producer<byte []> producer)
    {
        this.nodeType = new String(nodeType);
        this.queue = queue;
        this.consumer = consumer;
        this.apiURL = new String(apiURL);
        this.apiKey = new String(apiKey);
        this.producer = producer;
    }

    // 线程的 run 方法
    @Override
    public void run()
    {
        try
        {
            while(true)
            {
                // 尝试获取消息
                Message<byte []> msg = queue.take();
                try
                {
                    // 处理消息

                    // 如果是 Api 节点
                    if(nodeType.equals("Api"))
                    {
                        ApiMessageProcessor mp = new ApiMessageProcessor(apiURL,apiKey,producer);
                        mp.process(msg);
                    }
                    else if(nodeType.equals("GPU"))
                    {
                        
                    }
                    else
                    {
                        throw new IOException("The nodeType is illegal!\n");
                    }
                    // 处理成功 返回 ACK
                    consumer.acknowledge(msg);
                }
                // 处理失败 返回 NAK
                catch (Exception e)
                {
                    consumer.negativeAcknowledge(msg);
                }
            }
        }
        // 其他异常
        catch(Exception e)
        {
            e.printStackTrace();
        }
    }

    // 启动线程
    void start(String name) throws Exception
    {
        if(th == null)
        {
            th = new Thread(this,name);
            th.start();
        }
    }
}

File: WokerNodePython/WorkerNode/WorkerNode.py
from queue import Queue,Empty
from typing import Dict
import time
import os
import pulsar
import threading
import json
from ApiMessageProcessor import ApiMessageProcessor
import Event
from kubernetes import client,config

nodeType : str = "api";
pulsarURL : str = "pulsar://localhost:6650";
topicName : str = "model-gpt-3.5-turbo";
maxProcessNum : int = 10;
apiURL : str = "https://api.openai-hk.com/v1/chat/completions";
apiKey : str = "hk-j9e9al1000037138f0cd6a31058a83dbb7a63f56fd48788c";
queue : Queue = Queue();
map : Dict = dict();
model : str = 'gpt-3.5-turbo'
serviceTopicName : str = '' 
debug : bool = False
pulsarToken : str = ''
AIModelNamespace : str = ''
AIModelName : str = ''
apiInstance = None
stopEvent = threading.Event()
startTime = 0

def init():
    global nodeType,pulsarURL,serviceTopicName,pulsarToken,topicName
    global maxProcessNum,apiURL,apiKey,queue,map,model,debug,AIModelName,AIModelNamespace
    nodeType = os.getenv('NODE_TYPE','api');
    pulsarURL = os.getenv('PULSAR_URL',"pulsar://localhost:6650");
    maxProcessNum = int(os.getenv('MAX_PROCESS_NUM','128'));
    apiURL = os.getenv('API_URL',"https://api.openai-hk.com/v1/chat/completions");
    apiKey = os.getenv('API_KEY',"hk-j9e9al1000037138f0cd6a31058a83dbb7a63f56fd48788c");
    model = os.getenv('MODEL_NAME','gpt-3.5-turbo')
    serviceTopicName = os.getenv('RES_TOPIC_NAME','res-topic')
    debug = bool(os.getenv('DEBUG','false'))
    pulsarToken = os.getenv('PULSAR_TOKEN','')
    AIModelName = os.getenv('AIMODEL_NAME','none')
    AIModelNamespace = os.getenv('AIMODEL_NAMESPACE','none')
    topicName = 'model-' + model
    queue = Queue();
    map = dict();

def createConsumer(url):
    global pulsarClient
    pulsarClient = pulsar.Client(
        service_url=url,
        operation_timeout_seconds=10,
        )

    try:
        global consumer
        consumer = pulsarClient.subscribe(
            topic=topicName,
            subscription_name=topicName + '-subscription',
            consumer_type=pulsar.ConsumerType.Shared
            )
        return pulsarClient,consumer
    except Exception as e:
        Event.createEvent(apiInstance,AIModelName,AIModelNamespace,'MessageQueueConnectionError',e.__str__())
        exit(1)

def run():
    global pulsarClient,consumer

    pulsarClient,consumer = createConsumer(pulsarURL)
    
    processors = list()

    for i in range(maxProcessNum):
        processor = Processor('Thread-' + str(i),consumer)
        processor.start()
        processors.append(processor)

    try:
        while True:
            msg = consumer.receive()
            print('received message: {}'.format(msg.data())) # debug
            queue.put(msg,True)
    except KeyboardInterrupt:
        print('Stopping consumer...')
    finally:
        consumer.close()
        pulsarClient.close()


class Processor(threading.Thread):
    def __init__(self,name,consumer):
        threading.Thread.__init__(self)
        self.name = name
        self.consumer = consumer
        self.producer = pulsarClient.create_producer(serviceTopicName)
    def run(self):
        while True:
            msg = None
            if stopEvent.is_set() and time.time() > startTime:
                stopEvent.clear()
                global startTime
                startTime = 0
            elif stopEvent.is_set():
                time.sleep(5)
                continue
            try:
                msg = queue.get(True)
                print('{} take message: {}'.format(self.name,msg.data())) #debug
                amp = ApiMessageProcessor(msg,apiURL,apiKey,model)
                result = amp.process()
                self.sendResult(result)
                self.consumer.acknowledge(msg)
            except Empty:
                continue
            except json.JSONDecodeError as e:
                Event.createEvent(apiInstance,AIModelName,AIModelNamespace,'GeneralError','Error decoding JSON')
                if msg:
                    self.consumer.negative_acknowledge(msg)
            except Exception as e:
                if e.args[0] == 'Http request failed,status code: 428':
                    Event.createEvent(apiInstance,AIModelName,AIModelNamespace,'AuthenticationError','The key is wrong!')
                elif e.args[0] == 'Http request failed,status code: 429':
                    if not stopEvent.is_set():
                        stopEvent.set()
                        global startTime
                        startTime = time.time() + 60
                else:
                    Event.createEvent(apiInstance,AIModelName,AIModelNamespace,'GeneralError',e.__str__())
                if msg:
                    self.consumer.negative_acknowledge(msg)

    def sendResult(self,result):
        result['model'] = model
        result['ai_model_name'] = AIModelName
        result['ai_model_namespace'] = AIModelNamespace
        print('send result: ' + json.dumps(result))
        self.producer.send(json.dumps(result).encode('utf-8'))
        
                

def kubenetesInit():
    try:
        config.load_incluster_config()
    except:
        config.load_kube_config()
    global apiInstance
    apiInstance = client.CoreV1Api()
    
if __name__ == '__main__':
    kubenetesInit()
    init()
    run()